vng/vng_test.go

```diff
@@ -0,0 +1,351 @@
+		types := genTypes(bytesReader, context, 3)
+		values := genValues(bytesReader, context, types)
+		ColumnThresh := int(binary.LittleEndian.Uint64(genBytes(bytesReader, 8)))
+		if ColumnThresh == 0 {
+			ColumnThresh = 1
+		}
+		if ColumnThresh > vng.MaxSegmentThresh {
+			ColumnThresh = vng.MaxSegmentThresh
+		}
+		SkewThresh := int(binary.LittleEndian.Uint64(genBytes(bytesReader, 8)))
+		if SkewThresh == 0 {
+			SkewThresh = 1
+		}
+		if SkewThresh > vng.MaxSkewThresh {
+			SkewThresh = vng.MaxSkewThresh
+		}
+		writerOpts := vngio.WriterOpts{
+			ColumnThresh: units.Bytes(ColumnThresh),
+			SkewThresh:   units.Bytes(SkewThresh),
+		}
+		roundtrip(t, values, writerOpts)
+	})
+}
+
+func FuzzVngRoundtripBytes(f *testing.F) {
+	f.Fuzz(func(t *testing.T, b []byte) {
+		bytesReader := bytes.NewReader(b)
+		context := zed.NewContext()
+		reader := zngio.NewReader(context, bytesReader)
+		defer reader.Close()
+		var a zbuf.Array
+		if zio.Copy(&a, reader) != nil {
+			return
+		}
+		roundtrip(t, a.Values(), vngio.WriterOpts{
+			ColumnThresh: units.Bytes(vngio.DefaultColumnThresh),
+			SkewThresh:   units.Bytes(vngio.DefaultSkewThresh),
+		})
+	})
+}
+
+func roundtrip(t *testing.T, valuesIn []zed.Value, writerOpts vngio.WriterOpts) {
+	// Debug
+	//for i := range valuesIn {
+	//    t.Logf("value: in[%v].Bytes()=%v", i, valuesIn[i].Bytes())
+	//    t.Logf("value: in[%v]=%v", i, zson.String(&valuesIn[i]))
+	//}
+
+	// Write
+	var buf bytes.Buffer
+	writer, err := vngio.NewWriterWithOpts(zio.NopCloser(&buf), writerOpts)
+	require.NoError(t, err)
+	require.NoError(t, zio.Copy(writer, zbuf.NewArray(valuesIn)))
+	require.NoError(t, writer.Close())
+
+	// Read
+	reader, err := vngio.NewReader(zed.NewContext(), bytes.NewReader(buf.Bytes()))
+	require.NoError(t, err)
+	var a zbuf.Array
+	require.NoError(t, zio.Copy(&a, reader))
+	valuesOut := a.Values()
+
+	// Compare
+	t.Logf("comparing: len(in)=%v vs len(out)=%v", len(valuesIn), len(valuesOut))
+	for i := range valuesIn {
+		if i >= len(valuesOut) {
+			t.Errorf("missing value: in[%v].Bytes()=%v", i, valuesIn[i].Bytes())
+			t.Errorf("missing value: in[%v]=%v", i, zson.String(&valuesIn[i]))
+			continue
+		}
+		valueIn := valuesIn[i]
+		valueOut := valuesOut[i]
+		t.Logf("comparing: in[%v]=%v vs out[%v]=%v", i, zson.String(&valueIn), i, zson.String(&valueOut))
+		if !bytes.Equal(zed.EncodeTypeValue(valueIn.Type), zed.EncodeTypeValue(valueOut.Type)) {
+			t.Errorf("values have different types: %v vs %v", valueIn.Type, valueOut.Type)
+		}
+		if !bytes.Equal(valueIn.Bytes(), valueOut.Bytes()) {
+			t.Errorf("values have different zng bytes: %v vs %v", valueIn.Bytes(), valueOut.Bytes())
+		}
+	}
+	for i := range valuesOut[len(valuesIn):] {
+		t.Errorf("extra value: out[%v].Bytes()=%v", i, valuesOut[i].Bytes())
+		t.Errorf("extra value: out[%v]=%v", i, zson.String(&valuesOut[i]))
+	}
+}
+
+func genValues(b *bytes.Reader, context *zed.Context, types []zed.Type) []zed.Value {
+	values := make([]zed.Value, 0)
```

Comment:
We generally start with a nil slice (unless we want to return a zero-length slice instead of nil).
```suggestion
	var values []zed.Value
```